TELCODOCS#2230: Coordinating reboots for configuration changes (#91723)
openshift/openshift-docs: modules/ztp-coordinating-reboots-for-config-changes.adoc

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp_far_edge/ztp-configuring-managed-clusters-policies.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-coordinating-reboots-for-config-changes_{context}"]
= Coordinating reboots for configuration changes

You can use {cgu-operator-full} to coordinate reboots across a fleet of spoke clusters when configuration changes require a reboot.

The following types of configuration changes typically require a reboot:

* Updates to tuning profiles that modify kernel parameters or system behavior.
* Kubelet configuration updates.

> * Node-level changes delivered through `MachineConfig`, such as `sysctl` settings or system service changes.

Reviewer: I am not sure how accurate this is. Also I don't think this is necessary.
It can be just
You can use {cgu-operator-full} to coordinate reboots across a fleet of spoke clusters when configuration changes require a reboot, such as defered tuning changes.

Author: Updated as per the suggestion.

> {cgu-operator} ensures that only nodes with a degraded tuned profile, indicating a reboot is needed, are rebooted.

Reviewer: this is not true. TALM does not check for this. It reboot the nodes in the selcted mcp on the selected clusters.

Author: Can we update it as follows:

"TALM reboots all nodes in the targeted `MachineConfigPool` on the selected clusters when the reboot policy is applied."

Reviewer: yes, this correct

Instead of rebooting nodes after each individual change, you can apply all configuration updates through policies and then trigger a single, coordinated reboot. 

.Prerequisites

* You have installed the {oc-first}.
* You have logged in to the hub cluster as a user with `cluster-admin` privileges.
* You have deployed and configured {cgu-operator}.

.Procedure

. Create the configuration policies and policy bindings for the tuning or configuration changes.

. Create a reboot policy on the hub cluster. This policy checks for degraded tuned profiles that indicate a reboot is needed.

> . Create and apply the `ClusterGroupUpgrade` (CGU) custom resource (CR). In the `spec.managedPolicies` list, include all relevant configuration policies first, followed by the reboot policy and the optional `mcp-validator` policy.

Reviewer: The procedure should be mentioned:

- The policy can be generated using PolicyGenerator.
- Select `out/argocd/example/acmpolicygenerator/acm-example-sno-reboot` or `out/argocd/example/acmpolicygenerator/acm-example-multinode-reboot`.
- change `policyDefaults.placement.labelSelector` to match clusters that you want to reboot.
- modify other placeholder values according to your use case.
- in case of rebooting for applying defered tuning change, make sure mcp value matches Tuned.spec.recommended
- follow the procedure (Customizing a managed cluster with PolicyGenerator CRs)[link to section above in the same document]
- After ArgoCD syncs successfully, you can rollout the reboot using TALM by creating a CGU
example:

```
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  name: reboot
  namespace: default
spec:
  clusterLabelSelectors:
  - matchLabels:
      # select clusters that you want to reboot
  enable: true
  managedPolicies:
  # you add other configuration policies here, before the reboot policy
  - example-reboot
  remediationStrategy:
    timeout: 300 # timeout across all clusters; you should consider the worst case
    maxConcurrency: 10
 ```

Author: Updated as per the suggestion

+
After you apply the CGU custom resource, {cgu-operator} rolls out the configuration policies in order. When all policies are compliant, {cgu-operator} applies the reboot policy. This triggers a reboot of all nodes in the specified `MachineConfigPool`.

.Verification

> 

Reviewer: You can monitor the rollout on the hub by watching CGU's status. To verify all successful rollout of the reboot:
```
oc get cgu -A                                                                                                         
NAMESPACE   NAME     AGE   STATE       DETAILS
default     reboot   1d    Completed   All clusters are compliant with all the managed policies
```

Author: Updated as per the suggestion.

* Confirm that all nodes in the `MachineConfigPool` have rebooted.

> * Verify that the `MachineConfigPool` reaches the `Updated` state.

Reviewer: Verify successful reboot on a specific node:
```
oc get mcp master                                                                                                      +
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-be5785c3b98eb7a1ec902fef2b81e865   True      False      False      3              3                   3                     0                      72d
```

Author: Updated as per the suggestion.

// Module included in the following assemblies:
//

> // * scalability_and_performance/ztp_far_edge/ztp-configuring-managed-clusters-policies.adoc

Reviewer 2: Should this be `edge_computing/policygenerator_for_ztp/ztp-configuring-managed-clusters-policygenerator.adoc` ? because you added `modules/ztp-coordinating-reboots-for-config-changes.adoc` in that assembly?

So whichever is the correct assembly, enter the assembly's context in `ifeval` condition.

> 

Reviewer 2: ```suggestion

ifeval::["{context}" == "ztp-configuring-managed-clusters-policygenerator"]
:ztp:
endif::[]
```

Make sure to close this condition in the end of the file:

```
ifeval::["{context}" == "ztp-configuring-managed-clusters-policygenerator"]
:!ztp:
endif::[]
```

:_mod-docs-content-type: PROCEDURE

> [id="ztp-coordinating-reboots-for-config-changes_{context}"]
> = Coordinating reboots for configuration changes

Reviewer 2: ```suggestion
:_mod-docs-content-type: PROCEDURE
[id="ztp-coordinating-reboots-for-config-changes_{context}"]
ifndef::ztp[]
= Coordinating reboots for tuning profile changes
endif::ztp[]

ifdef::ztp[]
= Coordinating reboots for configuration changes
endif::ztp[]
```

You can use Topology Aware Lifecycle Manager (TALM) to coordinate reboots across a fleet of spoke clusters when configuration changes require a reboot.

> ifndef::ztp-configuring-managed-clusters-policygenerator[]
> These configuration changes include updates to tuning profiles that modify kernel parameters or system behavior.
> endif::[]

Reviewer 2: ```suggestion
ifndef::ztp[]
These configuration changes include updates to tuning profiles that modify kernel parameters or system behavior.
endif::ztp[]
```

Author: Hi @xenolinux 
There is another assembly where this file is included and that is - "using-node-tuning-operator.adoc"
And the above line must be visible only if this module is included in the node-tuning assembly.

Reviewer 2: You need one more ifeval for that assembly. Lets see he results then you can add more changes

> ifndef::node-tuning-operator[]
> The following types of configuration changes typically require a reboot:
> 
> * Updates to tuning profiles that modify kernel parameters or system behavior.node-tuning-operator
> * Kubelet configuration updates.
> * Node-level changes delivered through `MachineConfig`, such as `sysctl` settings or system service changes.
> endif::[]

Reviewer 2: ```suggestion
ifdef::ztp[]
The following types of configuration changes typically require a reboot:

* Updates to tuning profiles that modify kernel parameters or system behavior.node-tuning-operator
* Kubelet configuration updates.
* Node-level changes delivered through `MachineConfig`, such as `sysctl` settings or system service changes.
endif::ztp[]
```

> endif::[]

Reviewer 2: @sr1kar99 I added conditionals. Can you try these and see if the preview shows as expected?

// Module included in the following assemblies:
//
// * scalability_and_performance/ztp_far_edge/ztp-configuring-managed-clusters-policies.adoc
ifeval::["{context}" == "ztp-configuring-managed-clusters-policygenerator"]
:ztp:
endif::[]

:_mod-docs-content-type: PROCEDURE
[id="ztp-coordinating-reboots-for-config-changes_{context}"]
ifdef::ztp[]
= Coordinating reboots for configuration changes
endif::[]
ifndef::ztp[]
= Coordinating reboots for tuned profile changes

> endif::[]

Reviewer 2: ```suggestion
endif::ztp[]
```

Look out for other endifs where ztp is missing

. Apply the `PolicyGenerator` CR to generate and apply the configuration policies. For detailed steps, see "Customizing a managed cluster with PolicyGenerator CRs".

. After ArgoCD completes syncing the policies, create and apply the `ClusterGroupUpgrade` (CGU) CR. 
+
.Example CGU custom resource configuration
[source,yaml]
----
apiVersion: ran.openshift.io/v1alpha1
kind: ClusterGroupUpgrade
metadata:
  name: reboot
  namespace: default
spec:
  clusterLabelSelectors:
  - matchLabels: <1>
# ...
  enable: true
  managedPolicies: <2>
  - example-reboot
  remediationStrategy:
    timeout: 300 <3>
    maxConcurrency: 10
# ...
----
<1> Configure the labels that match the clusters you want to reboot.
<2> Add all required configuration policies before the reboot policy. {cgu-operator} applies the configuration changes as specified in the policies, in the order they are listed.
<3> Specify the timeout in seconds for the entire upgrade across all selected clusters. Set this field by considering the worst-case scenario.

> . After you apply the CGU custom resource, {cgu-operator} rolls out the configuration policies in order. Once all policies are compliant, it applies the reboot policy and triggers a reboot of all nodes in the specified `MachineConfigPool`.

Reviewer 3: This isn't really a step that the customer performs. This is a consequence of applying the CR. 

This is not a merge blocker, but in a follow up PR, I recommend explicitly adding the `oc apply -f` command and then you can include this information below the procedure step, similar to the following example:
```suggestion
. Apply the CGU CR by entering the following command:
+
[source,terminal]
----
$ oc apply -f <cgu_cr>.yaml
----
+
After you apply the CGU custom resource, {cgu-operator} rolls out the configuration policies in order. Once all policies are compliant, it applies the reboot policy and triggers a reboot of all nodes in the specified machine config pool.
```
Also, when you use a generic reference to an API object, as you do in the previous sentence, the guidance is to drop the camel case and backticks, hence "machine config pool". If you are referring to the specific API object, which you can do but I don't think is necessary, then `MachineConfigPool` should be followed by a noun such as _object_ or _resource_. See the repo guidelines for more information: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting

.Verification

. Monitor the CGU rollout status.
+
You can monitor the rollout of the CGU custom resource on the hub by checking the status. Verify the successful rollout of the reboot by running the following command:
+
[source,terminal]
----
oc get cgu -A
----
+
.Example output
[source,terminal]
----
NAMESPACE   NAME     AGE   STATE       DETAILS
default     reboot   1d    Completed   All clusters are compliant with all the managed policies
----

. Verify successful reboot on a specific node.
+

> To confirm that the reboot was successful on a specific node, check the status of the `MachineConfigPool` (MCP) for the node by running the following command:

Reviewer 3: Similar to the comment above, to conform with the rules around API objects, the correct usage is "`MachineConfigPool` <noun>" or simply "machine config pool". Not a merge blocker.
